### ISSUE TITLE
Permission denied on msal_token_cache.bin.lockfile

### ISSUE THREAD

**Issue by Stolpe** [Azure/azure-cli#20273]
## Describe the bug
When running Terraform locally, I get an error originating from the Azure CLI. This happens on several versions, all relatively new, of the azurerm Terraform module. I am able to reproduce the error without going via Terraform, so I think that this issue ought to be filed in this repo.

**Command Name**
`az account get-access-token`

**Errors:**
```
ERROR: The command failed with an unexpected error. Here is the traceback:
ERROR: [Errno 13] Permission denied: 'C:\\Users\\Per Stolpe\\.azure\\msal_token_cache.bin.lockfile'
Traceback (most recent call last):
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\knack/cli.py", line 231, in invoke
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 657, in execute
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 720, in _run_jobs_serially
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 691, in _run_job
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 328, in __call__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/command_operation.py", line 121, in handler
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/command_modules/profile/custom.py", line 74, in get_access_token      
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/_profile.py", line 381, in get_raw_token
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/_profile.py", line 588, in _create_credential
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/auth/identity.py", line 182, in get_user_credential
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/auth/msal_authentication.py", line 41, in __init__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\msal/application.py", line 872, in get_accounts
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\msal/application.py", line 910, in _find_msal_accounts
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\msal_extensions/token_cache.py", line 53, in find
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\msal_extensions/cache_lock.py", line 29, in __enter__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\portalocker/utils.py", line 199, in __enter__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\portalocker/utils.py", line 128, in acquire
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\portalocker/utils.py", line 175, in _get_fh
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\Per Stolpe\\.azure\\msal_token_cache.bin.lockfile'
```

## To Reproduce:
The azurerm Terraform module fires off several commands in parallel, triggering this error. Since this resembles a race condition, I run below Powershell and wait until it says that something went wrong.
```powershell
0..1000 | `
    Foreach-Object -Parallel {
        $subscriptionId = switch ($_ % 9) {
            0 { 'subscription id guid'; break }
            1 { 'subscription id guid'; break }
            2 { 'subscription id guid'; break }
            3 { 'subscription id guid'; break }
            4 { 'subscription id guid'; break }
            5 { 'subscription id guid'; break }
            6 { 'subscription id guid'; break }
            7 { 'subscription id guid'; break }
            8 { 'subscription id guid'; break }
        }

        az account get-access-token --subscription $subscriptionId
    }
```

## Expected Behavior
Just the access token JSON printed a thousand times.

## Environment Summary
```
Windows-10-10.0.19041-SP0
Python 3.8.9
Installer: MSI

azure-cli 2.30.0

```
## Additional Context
When I first tried to recreate this issue using only four subscription ids, it never happened. Thus, it seems that an increasing amount of subscription ids increases the probability of triggering this issue. In my Terraform code, I call multiple Terraform modules, which themselves call other modules, so nine subscription ids is a realistic amount for me at least.

I should add that I have tried to reboot, uninstall and reinstall the CLI.

**Comment by yonzhan** [comment]
@jiasli for awareness

**Comment by Stolpe** [comment]
Note that this error is not triggered for me every time i run this script, but here is some more output from another run which triggered the error. The output is preceded by only successful outputs, but the list of recent commands lists multiple failures. This might, however, be due to me aborting the run using `Ctrl` + `c`. I thought it was interesting enough to mention, though.
```
{
  "accessToken": "ey[redacted]",
  "expiresOn": "2021-11-10 16:21:32.000000",
  "subscription": "[redacted]",
  "tenant": "[redacted]",
  "tokenType": "Bearer"
}
{
  "accessToken": "ey[redacted]",
  "expiresOn": "2021-11-10 16:21:32.000000",
  "subscription": "[redacted]",
  "tenant": "[redacted]",
  "tokenType": "Bearer"
}
ERROR: The command failed with an unexpected error. Here is the traceback:
ERROR: [Errno 13] Permission denied: 'C:\\Users\\Per Stolpe\\.azure\\msal_token_cache.bin.lockfile'
Traceback (most recent call last):
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\knack/cli.py", line 231, in invoke
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 657, in execute
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 720, in _run_jobs_serially
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 691, in _run_job
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 328, in __call__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/command_operation.py", line 121, in handler
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/command_modules/profile/custom.py", line 74, in get_access_token
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/_profile.py", line 381, in get_raw_token
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/_profile.py", line 588, in _create_credential
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/auth/identity.py", line 182, in get_user_credential
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/auth/msal_authentication.py", line 41, in __init__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\msal/application.py", line 872, in get_accounts
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\msal/application.py", line 910, in _find_msal_accounts
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\msal_extensions/token_cache.py", line 53, in find
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\msal_extensions/cache_lock.py", line 29, in __enter__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\portalocker/utils.py", line 199, in __enter__
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\portalocker/utils.py", line 128, in acquire
  File "D:\a\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\portalocker/utils.py", line 175, in _get_fh
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\Per Stolpe\\.azure\\msal_token_cache.bin.lockfile'
To open an issue, please run: 'az feedback'
{
  "accessToken": "ey[redacted]",
  "expiresOn": "2021-11-10 16:21:32.000000",
  "subscription": "[redacted]",
  "tenant": "[redacted]",
  "tokenType": "Bearer"
}
PS C:\> az feedback

We appreciate your feedback!

For more information on getting started, visit: aka.ms/azcli/get-started
If you have questions, visit our Stack Overflow page: aka.ms/azcli/questions
Please let us know how we are doing: https://aka.ms/azureclihats

Recent commands:

   [0] create a generic issue.
   [1] az account get-access-token: SUCCESS. Ran: 11 secs ago.
   [2] az account get-access-token: FAILURE. Ran: 11 secs ago.
   [3] az account get-access-token: SUCCESS. Ran: 11 secs ago.
   [4] az account get-access-token: SUCCESS. Ran: 11 secs ago.
   [5] az account get-access-token: FAILURE. Ran: 9 secs ago.
   [6] az account get-access-token: FAILURE. Ran: 9 secs ago.
   [7] az account get-access-token: FAILURE. Ran: 9 secs ago.
   [8] az account get-access-token: FAILURE. Ran: 9 secs ago.
   [9] az account get-access-token: FAILURE. Ran: 9 secs ago.

Enter the number of the command you would like to create an issue for. Enter q to quit:
```

**Comment by jiasli** [comment]
I added a `sleep` statement in `msal_extensions.token_cache.PersistedTokenCache.find`:

```py
    def find(self, credential_type, **kwargs):  # pylint: disable=arguments-differ
        with CrossPlatLock(self._lock_location):
            print("sleeping")
            import time
            time.sleep(30)
            self._reload_if_necessary()
            return super(PersistedTokenCache, self).find(credential_type, **kwargs)
```

Then run the PowerShell script:

```powershell
0..1000 | `
    Foreach-Object -Parallel {
        az account get-access-token
    }
```

I got error:

```
  File "D:\cli\py310\lib\site-packages\msal_extensions\token_cache.py", line 53, in find
    with CrossPlatLock(self._lock_location):
  File "D:\cli\py310\lib\site-packages\msal_extensions\cache_lock.py", line 29, in __enter__
    file_handle = self._lock.__enter__()
  File "D:\cli\py310\lib\site-packages\portalocker\utils.py", line 199, in __enter__
    return self.acquire()
  File "D:\cli\py310\lib\site-packages\portalocker\utils.py", line 158, in acquire
    raise exceptions.LockException(exception)
portalocker.exceptions.LockException: (1, 'Permission denied')
```

Even though the error is not exactly the same as yours, I think this is because `msal-extension` sets `portalocker.LOCK_EX | portalocker.LOCK_NB`.

https://github.com/AzureAD/microsoft-authentication-extensions-for-python/blob/30b3e867b88bb73908e681719bd312864f69772e/msal_extensions/cache_lock.py#L33

```py
        self._lock = portalocker.Lock(
            lockfile_path,
            ...
            flags=portalocker.LOCK_EX | portalocker.LOCK_NB,
```

This makes MSAL fail on concurrent execution.

I am still trying to reproduce and investigate your error message.

A similar topic was discussed before at https://github.com/Azure/azure-cli/issues/9427. For now, the best workaround is to avoid reading/writing the same `~/.azure` folder concurrently by setting `AZURE_CONFIG_DIR` environment variable https://github.com/Azure/azure-cli/issues/9427#issuecomment-647865668.

**Comment by jiasli** [comment]
I am able to repro this error _only once_:

```
PermissionError: [Errno 13] Permission denied: 'C:\\Users\\myname\\.azure\\msal_token_cache.bin.lockfile'
```

Later it looks like 

```
portalocker.exceptions.LockException: (1, 'Permission denied')
```

**Comment by jiasli** [comment]
## In ADAL-based Azure CLI

Previously (`azure-cli<=2.29.2`), there is no lock at all, which may cause problem only during **concurrent writes** to `~/.azure/accessTokens.json` for token refreshing (https://github.com/Azure/azure-cli/issues/9427). **Concurrent writes** only happen once an hour, as the access token expires once an hour. 

Besides this, **concurrent reads** work as expected.

## In MSAL-based Azure CLI

But now, as we can see from `msal-extension`'s code:

https://github.com/AzureAD/microsoft-authentication-extensions-for-python/blob/8605a9503aab930397efbbd5caea5d19b2713ced/msal_extensions/token_cache.py#L52-L55

```py
    def find(self, credential_type, **kwargs):  # pylint: disable=arguments-differ
        with CrossPlatLock(self._lock_location):
            self._reload_if_necessary()
            return super(PersistedTokenCache, self).find(credential_type, **kwargs)
```

even read requires a lock. This breaks **concurrent reads** and significantly worsens Azure CLI's concurrency support.

**Comment by Stolpe** [comment]
Thank you for your elaborate answers and for you work. I can confirm that my use-case (IE Terraform using the azurerm module) works with version 2.29.2. I will raise a ticket at their repo to let them know this is happening. Is there anything else you want for me to do?

**Comment by jiasli** [comment]
@Stolpe for the update. I can't repro your exact error message. Is there anything else I am missing?

**Comment by Stolpe** [comment]
No. I get the error message perhaps once every five tries or so. Perhaps something induces a wait on my machine that does not happen on your machine, antivirus or whatever.

What _does_ trigger my error message _every_ time is running a `terraform plan` with my setup of modules and resources. While I'm not able to share that setup, I did create a minimal(?) repro at https://github.com/Stolpe/terraform-azurecli-repro. I fully realize that neither Terraform nor the azurerm module is your responsibility, but perhaps you are able to see exactly what happens when it calls the CLI. I tested this with the newly released version 1.0.11 version of Terraform on Windows.

To trigger the error, modify `import.ps1` to point to real resource groups in two different subscriptions. Then run `terraform plan`. This will only import the existing infrastructure into a local Terraform state file, and then poll them. This does not modify your infrastructure at all.

**Comment by jiasli** [comment]
@Stolpe we are working on https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/100 to refine concurrent read access of Azure CLI and MSAL.

Perhaps you can help install that feature branch `skip-read-lock` of `msal-extensions` and see if the issue is mitigated?

Open a PowerShell terminal with **Administrator** permission, then run:

```
& "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\python.exe" -m pip uninstall --yes msal-extensions
& "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\python.exe" -m pip install git+https://github.com/AzureAD/microsoft-authentication-extensions-for-python@skip-read-lock#egg=msal-extensions
```

This will install the feature branch `skip-read-lock` of `msal-extensions` to `C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\Lib\site-packages\msal_extensions`.

**Comment by Stolpe** [comment]
Hi @jiasli,

I installed the feature branch according to your instructions above, and that did indeed solve the issue I had. Admittedly, the issue was far from easily reproducible with my repro repo, but with my Terraform configuration, it happened every time. Now, that too works flawlessly with version 2.30.0 of the Azure CLI.

Thank you very much for your help. You may close this issue as resolved when the time is right for that.

**Comment by rayluo** [comment]
> Hi @jiasli,
> 
> I installed the feature branch according to your instructions above, and that did indeed solve the issue I had. Admittedly, the issue was far from easily reproducible with my repro repo, but with my Terraform configuration, it happened every time. Now, that too works flawlessly with version 2.30.0 of the Azure CLI.
> 
> Thank you very much for your help. You may close this issue as resolved when the time is right for that.

@stolpe, thanks for sharing the good news, so far.

In one of your earlier message, you mentioned something like this:

```
{
  "accessToken": "ey[redacted]",
  "expiresOn": "2021-11-10 16:21:32.000000",
  "subscription": "[redacted]",
  "tenant": "[redacted]",
  "tokenType": "Bearer"
}
```

When was that `expiresOn` time? Assuming it was one hour after that token was issued, I would suggest you keep an eye on your Terraform setup for one hour plus several minutes, and see if the tokens would be refreshed successfully. Please report back to us. Thanks!

**Comment by Stolpe** [comment]
@rayluo, `expiresOn` seems to be set two hours ahead for me, and it automatically refreshes using version 2.30.0 of the CLI and the feature branch of MSAL, described above.

**Comment by rayluo** [comment]
Hi @stolpe, if you won't mind, would you please help us try one more thing? Uninstall the msal-extensions again, and then reinstall its dev branch, which is a branch does NOT yet contain our recent adjustment. The reason is that we happen to have 3 different versions of msal-extensions: (1) one that being released and used by Az CLI, (2) one in the `dev` branch with some lock adjustment that was not shipped, and then (3) the latest `skip-read-lock` branch.

The following commands installs the `dev` branch. Please let us know whether that also helps. We want to understand whether that could also make a difference. (And if not, you can always redo @jiasli 's earlier commands to switch back to the latest `skip-read-lock` branch.)

```
& "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\python.exe" -m pip uninstall --yes msal-extensions
& "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\python.exe" -m pip install git+https://github.com/AzureAD/microsoft-authentication-extensions-for-python@dev#egg=msal-extensions
```

Thanks for your understanding and contribution!

**Comment by Stolpe** [comment]
Hi @rayluo,

When I installed the dev branch as instructed above, the error reappeared for me.

(I always had the workaround of using 2.29 of the CLI, so no worries about my environment)

**Comment by kensykora** [comment]
I'm also experiencing this issue. Downgrading to 2.29.2 fixed the issue for me.

```bash
sudo apt-get install azure-cli=2.29.2-1~focal
```

**Comment by devbeard** [comment]
In addition to downgrading to make terraform work again, I also had to set `az config set auto-upgrade.enable=no` so the run in terraform could continue unprompted.

**Comment by rayluo** [comment]
> I'm also experiencing this issue. Downgrading to 2.29.2 fixed the issue for me.
> 
> ```shell
> sudo apt-get install azure-cli=2.29.2-1~focal
> ```



> In addition to downgrading to make terraform work again, I also had to set `az config set auto-upgrade.enable=no` so the run in terraform could continue unprompted.

Actually a better workaround is to keep using latest azure-cli 2.30 but manually upgrade one of its dependency by:

```
& "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\python.exe" -m pip uninstall --yes msal-extensions
& "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\python.exe" -m pip install git+https://github.com/AzureAD/microsoft-authentication-extensions-for-python@dev#egg=msal-extensions
```

**Comment by rayluo** [comment]
@stolpe, @kensykora , @devbeard , thanks for your patience. The issue has been fixed in Azure CLI 2.30+ 's upstream module, `msal-extensions`. Now, users of Azure CLI 2.30 and 2.31 can simply upgrade your environment like this:

    "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\python.exe" -m pip install -U --yes msal-extensions

We expect upcoming Azure CLI 2.32 will include this module by default.

(@jiasli , feel free to convert this issue to a Q&A, and mark this message as the final answer.)

**Comment by jiasli** [comment]
This issue is resolved by https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/100, please update to the latest Azure CLI.

**Comment by Stolpe** [comment]
I can confirm that the updated version of the Azure CLI works perfectly for my "real" use-case with Terraform, which caused me to create this issue. Thank you and good job!

**Comment by sanistaal-bbc** [comment]
I just installed a new VM with Azure CLI version 2.34.1 and I get the error occasionally now :-( Can I downgrade to a specific version to fix this?

**Comment by rayluo** [comment]
> I just installed a new VM with Azure CLI version 2.34.1 and I get the error occasionally now :-( Can I downgrade to a specific version to fix this?

Can you share more on your usage scenario, such as: what kind of Azure CLI commands that you use, what is the VM's OS, and the error trace you got, etc.? That could help us investigate and try to find the root cause.

By the way, you may consider create a new github issue for your case. I'm not sure whether interactions on a closed issue would still grab enough attention from @jiasli's team.

**Comment by sanistaal-bbc** [comment]
> > I just installed a new VM with Azure CLI version 2.34.1 and I get the error occasionally now :-( Can I downgrade to a specific version to fix this?
> 
> Can you share more on your usage scenario, such as: what kind of Azure CLI commands that you use, what is the VM's OS, and the error trace you got, etc.? That could help us investigate and try to find the root cause.
> 
> By the way, you may consider create a new github issue for your case. I'm not sure whether interactions on a closed issue would still grab enough attention from @jiasli's team.

You are right about creating a new issue. I will do that and thanks
